File: packages/mobile/otp-utils/ios/Sources/OTPUtils/OtpValidation.swift
import Foundation

public let otpLength: Int = 6
public let resendIntervalSeconds: TimeInterval = 30

public func sanitisePhoneNumber(_ raw: String) -> String {
    let allowed = raw.replacingOccurrences(of: "[^0-9+]", with: "", options: .regularExpression)
    let singlePlus = allowed.replacingOccurrences(of: "\\+{2,}", with: "+", options: .regularExpression)
    return singlePlus.replacingOccurrences(of: "(\\+)(?=\\+)", with: "", options: .regularExpression)
}

public func normalisePhoneNumber(_ raw: String) -> String {
    let trimmed = sanitisePhoneNumber(raw).trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed.hasPrefix("+") {
        return trimmed
    }
    return "+" + trimmed
}

public func normaliseOtp(_ raw: String) -> String {
    let digits = raw.replacingOccurrences(of: "\\D", with: "", options: .regularExpression)
    return String(digits.prefix(otpLength))
}

public func isValidOtp(_ raw: String) -> Bool {
    return normaliseOtp(raw).count == otpLength
}

public final class ResendController {
    private let windowSeconds: TimeInterval
    private var nextAllowedAt: Date = .distantPast

    public init(windowSeconds: TimeInterval = resendIntervalSeconds) {
        self.windowSeconds = windowSeconds
    }

    public func canSend(now: Date = .init()) -> Bool {
        return now >= nextAllowedAt
    }

    public func registerSend(now: Date = .init()) {
        nextAllowedAt = now.addingTimeInterval(windowSeconds)
    }

    public func secondsUntilNext(now: Date = .init()) -> Int {
        guard !canSend(now: now) else { return 0 }
        return Int(ceil(nextAllowedAt.timeIntervalSince(now)))
    }
}

File: packages/mobile/otp-utils/android/OtpValidation.kt
package com.rayon.mobile.otputils

import kotlin.math.ceil

const val OTP_LENGTH = 6
const val RESEND_INTERVAL_SECONDS = 30

fun sanitisePhoneNumber(raw: String): String =
  raw.replace(Regex("[^0-9+]"), "")
    .replace(Regex("\\+{2,}"), "+")
    .replace(Regex("(\\+)(?=\\+)"), "")

fun normalisePhoneNumber(raw: String): String {
  val trimmed = sanitisePhoneNumber(raw).trim()
  return if (trimmed.startsWith("+")) trimmed else "+$trimmed"
}

fun normaliseOtp(raw: String): String = raw.replace(Regex("\\D"), "").take(OTP_LENGTH)

fun isValidOtp(raw: String): Boolean = normaliseOtp(raw).length == OTP_LENGTH

class ResendController(private val windowSeconds: Int = RESEND_INTERVAL_SECONDS) {
  private var nextAllowedAt: Long = 0

  fun canSend(now: Long = System.currentTimeMillis()): Boolean = now >= nextAllowedAt

  fun registerSend(now: Long = System.currentTimeMillis()) {
    nextAllowedAt = now + windowSeconds * 1_000L
  }

  fun secondsUntilNext(now: Long = System.currentTimeMillis()): Int {
    if (canSend(now)) return 0
    return ceil((nextAllowedAt - now) / 1_000.0).toInt()
  }
}
